buildSrc/subprojects/parser/src/main/kotlin/org/apache/calcite/buildtools/parser/DialectGenerateTask.kt

```diff
@@ -0,0 +1,210 @@
+     * <return_type> <name>(<args>) :
+     * {
+     *     <properties>
+     * }
+     * {
+     *     <body>
+     * }
+     * @param file The file to process
+     * @param functionMap The map to which the parsing functions will be added to
+     */
+    private fun processFile(file: File, functionMap: MutableMap<String, String>) {
+        var fileText = file.readText(Charsets.UTF_8)
+        val typeAndName = "\\w+\\s+\\w+"
+        val declarationPattern =
+            Regex("(%s\\s*\\(\\s*(%s\\s*(\\,\\s*%s\\s*)*)?\\)\\s*\\:\n)"
+                    .format(typeAndName, typeAndName, typeAndName))
+        val delims = "(\\s|\n|\"|(//)|(/\\*)|(\\*/))"
+        // Uses Lookahead and Lookbehind to ensure that the delims are added as tokens.
+        val splitRegex = Regex("((?<=%s)|(?=%s))".format(delims, delims))
+        val declarationMatches = declarationPattern.findAll(fileText)
+        for (m in declarationMatches) {
+            val functionDeclaration = m.value
+            val functionName = getFunctionName(functionDeclaration)
+            val functionBuilder = StringBuilder(functionDeclaration)
+            val declarationIndex = fileText.indexOf(functionDeclaration)
+            fileText = fileText.substring(declarationIndex + functionDeclaration.length)
+            val tokens: Queue<String> = LinkedList(splitRegex.split(fileText))
+            processCurlyBlock(functionBuilder, tokens)
+            processCurlyBlock(functionBuilder, tokens)
+            functionMap.put(functionName, functionBuilder.toString())
+        }
+    }
+
+    /**
+     * Parses a block of text surrounded by curly braces. The function keeps track
+     * of the number of curly braces that have yet to be closed. Once this counter
+     * reaches 0 the block has been fully parsed. The function also ensures that
+     * curly braces within single-line comments, multi-line comments, and strings
+     * are not counted for this.
+     *
+     * @param functionBuilder The builder unto which the tokens get added to once parsed
+     * @param tokens The tokens to be parsed
+     */
+    private fun processCurlyBlock(functionBuilder: StringBuilder, tokens: Queue<String>) {
+        var curlyCounter = 0
+        var insideString = false
+        var insideSingleLineComment = false
+        var insideMultiLineComment = false
+        var validBrace = true
+        while (tokens.isNotEmpty()) {
+            val token = tokens.poll()
+            functionBuilder.append(token)
+            validBrace = !insideString &&
+                    !insideSingleLineComment &&
+                    !insideMultiLineComment
+            if (token == "\n") {
+                if (insideSingleLineComment) {
+                    insideSingleLineComment = false
+                }
+                // Allows for an abitrary amount of new lines to be parsed.
+                continue
+            }
+            if (token == "\"" && !insideSingleLineComment && !insideMultiLineComment) {
```

Comment:
Can strings be marked by any other characters besides a double quote? Is a single quote valid?

Reply:
I think single quotes are used for characters, I can add a similar case for it.